### ISSUE TITLE
Kserve Co-Deployment: Inference + Indexed Data Integration

### ISSUE THREAD

**Issue by rch** [rch/asf-kudu#12]
## Overview
Deploy Kserve (CNCF incubating project) alongside Apache Pinot on Kubernetes to provide a unified platform for ML model serving with low-latency access to indexed actor state data, enabling feature serving, RAG pipelines, and contextual inference.

**Parent Issue**: #8 (Apache Pinot Integration for Indexed Materialized Views)

## Motivation
- **Feature serving**: Pinot provides <10ms p99 feature lookups for model inference
- **RAG pipelines**: Vector search in Pinot powers semantic retrieval for LLMs
- **Contextual inference**: Enrich predictions with real-time actor state context
- **Unified platform**: Single K8s deployment for analytics + ML serving
- **Cost efficiency**: Shared infrastructure and data locality

## Kserve Overview (2025 CNCF Incubating Project)

**Status**: Kserve moved to CNCF as an incubating project in September 2025

**Key Features**:
- Multi-framework support: TensorFlow, PyTorch, XGBoost, scikit-learn, ONNX, LLMs
- Serverless deployment with Knative autoscaling (scale-to-zero)
- KEDA integration for LLM autoscaling (HPA extension)
- Model serving protocol: V1 (TensorFlow) and V2 (KServe standard)
- Advanced features: canary rollouts, A/B testing, model monitoring

### Kubernetes Namespace Layout
```
Namespace: caf-ml-analytics
├── Apache Pinot Cluster
│   ├── Pinot Controller (StatefulSet)
│   ├── Pinot Broker (Deployment)
│   ├── Pinot Server (StatefulSet) - with indexes
│   └── Pinot Minion (Deployment) - optional
├── Kserve InferenceServices
│   ├── Feature Enrichment Transformer
│   ├── Embedding Generation Service
│   ├── Anomaly Detection Model
│   └── Contextual Recommendation Model
├── Shared Infrastructure
│   ├── Flink/Kafka (stream processing and event streaming)
│   ├── Ceph/S3 (model store, Pinot deep store)
│   ├── PostgreSQL (metadata, catalogs)
│   └── Redis (caching layer)
└── Monitoring Stack
    ├── Prometheus
    ├── Grafana
    └── Jaeger (distributed tracing)
```

#### Pattern 1: Feature Serving for Inference
```
Inference Request
  → Kserve Transformer (feature extraction)
  → Pinot Query (fetch features via star-tree index)
  → Kserve Predictor (model inference)
  → Response with prediction + metadata
```

#### Pattern 2: Embedding Generation Pipeline
```
Actor State Text
  → Kserve InferenceService (embedding model)
  → Generate embedding vector (768-dim)
  → Kafka Topic (embeddings)
  → Pinot Ingestion (vector index)
```

#### Pattern 3: RAG-Based Contextual Inference
```
User Query
  → Embedding Service (query → vector)
  → Pinot Vector Search (find relevant contexts)
  → LLM Prompt Construction (context injection)
  → Kserve LLM Inference (GPT/Llama/etc)
  → Response with citations
```

### 1. Real-Time Feature Serving
Serve features from Pinot for online ML predictions:

**Kserve Transformer** (pre-processing):
```python
class FeatureTransformer(kserve.Model):
    def preprocess(self, inputs):
        actor_id = inputs["actor_id"]

        # Query Pinot for real-time features
        features = pinot_client.query(f"""
            SELECT cpu_usage_pct, memory_mb, events_processed_1h,
                   error_rate, avg_latency_ms
            FROM actor_metrics
            WHERE actor_id = '{actor_id}'
              AND timestamp > now() - INTERVAL '5' MINUTE
            ORDER BY timestamp DESC
            LIMIT 1
        """)

        return {"features": features[0]}
```

**Kserve Predictor** (model inference):
```python
class AnomalyPredictor(kserve.Model):
    def predict(self, request):
        features = request["features"]
        prediction = self.model.predict([features])
        return {"anomaly_score": float(prediction[0])}
```

### 2. Embedding Generation Service
Generate embeddings for vector search in Pinot:

**InferenceService YAML**:
```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: embedding-generator
spec:
  predictor:
    containers:
    - name: embedder
      image: sentence-transformers/all-MiniLM-L6-v2
      env:
      - name: MODEL_NAME
        value: all-MiniLM-L6-v2
      - name: OUTPUT_KAFKA_TOPIC
        value: actor_embeddings
```

**Client Code**:
```python
# Generate embedding via Kserve
response = requests.post(
    "http://embedding-generator.caf-ml-analytics.svc.cluster.local/v1/models/embedder:predict",
    json={"instances": [{"text": "Actor processing payment in retry state"}]}
)
embedding = response.json()["predictions"][0]

# Query similar actors in Pinot
similar_actors = pinot_client.query(f"""
    SELECT actor_id, state_description,
           VECTOR_SIMILARITY(state_embedding, ARRAY{embedding}, 'COSINE') as similarity
    FROM actor_embeddings
    ORDER BY similarity DESC
    LIMIT 10
""")
```

### 3. RAG Pipeline for Actor Insights
Use Pinot vector search for retrieval-augmented generation:

**Transformer (retrieval)**:
```python
class RAGTransformer(kserve.Model):
    def preprocess(self, inputs):
        query = inputs["question"]

        # Generate query embedding via Kserve embedding service
        query_embedding = self.embedding_service.predict(query)

        # Retrieve relevant contexts from Pinot vector index
        contexts = pinot_client.query(f"""
            SELECT context_text, actor_id, metadata,
                   VECTOR_SIMILARITY(context_embedding, ARRAY{query_embedding}, 'COSINE') as relevance
            FROM actor_contexts
            ORDER BY relevance DESC
            LIMIT 5
        """)

        # Construct LLM prompt with retrieved context
        prompt = f"Question: {query}\n\nContext:\n"
        for ctx in contexts:
            prompt += f"- {ctx['context_text']} (actor: {ctx['actor_id']})\n"

        return {"prompt": prompt, "contexts": contexts}
```

**Predictor (LLM inference)**:
```python
class LLMPredictor(kserve.Model):
    def predict(self, request):
        prompt = request["prompt"]
        response = self.llm.generate(prompt, max_tokens=500)
        return {
            "answer": response,
            "sources": request["contexts"]
        }
```

### 4. Geospatial-Enhanced Inference
Enrich predictions with location context from Pinot:

```python
class LocationAwarePredictor(kserve.Model):
    def preprocess(self, inputs):
        lon, lat = inputs["longitude"], inputs["latitude"]

        # Find nearby actors using Pinot geospatial index
        nearby_actors = pinot_client.query(f"""
            SELECT actor_id, actor_type, status,
                   ST_DISTANCE(location, ST_Point({lon}, {lat}, 1)) as distance_m
            FROM actor_locations
            WHERE ST_DISTANCE(location, ST_Point({lon}, {lat}, 1)) < 1000
            ORDER BY distance_m
            LIMIT 10
        """)

        # Aggregate nearby context
        context = {
            "nearby_count": len(nearby_actors),
            "types_distribution": ...,
            "avg_health_score": ...
        }

        return {**inputs, "location_context": context}
```

### Pinot StatefulSet (with indexed data)
```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: pinot-server
spec:
  serviceName: pinot-server
  replicas: 3
  template:
    spec:
      containers:
      - name: pinot-server
        image: apachepinot/pinot:latest
        resources:
          requests:
            memory: "16Gi"
            cpu: "4"
          limits:
            memory: "32Gi"
            cpu: "8"
        volumeMounts:
        - name: data
          mountPath: /var/pinot/server/data
  volumeClaimTemplates:
  - metadata:
      name: data
    spec:
      accessModes: ["ReadWriteOnce"]
      resources:
        requests:
          storage: 500Gi
```

### Kserve InferenceService
```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: actor-anomaly-detector
  namespace: caf-ml-analytics
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      storageUri: s3://models/anomaly-detector/v1
      resources:
        requests:
          memory: 2Gi
          cpu: 1
        limits:
          memory: 4Gi
          cpu: 2
  transformer:
    containers:
    - name: feature-enrichment
      image: my-registry/pinot-feature-transformer:v1
      env:
      - name: PINOT_BROKER_URL
        value: "http://pinot-broker.caf-ml-analytics.svc.cluster.local:8099"
```

### Service Mesh Integration (Optional)
Use Istio for traffic management and observability:
```yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: ml-inference
spec:
  hosts:
  - ml-inference.example.com
  http:
  - match:
    - uri:
        prefix: /v1/models/anomaly
    route:
    - destination:
        host: actor-anomaly-detector-predictor-default
      weight: 90
    - destination:
        host: actor-anomaly-detector-predictor-canary
      weight: 10
```

## Implementation Tasks
- [ ] Design Kubernetes namespace architecture
- [ ] Deploy Pinot cluster on K8s (Helm or Kustomize)
- [ ] Deploy Kserve control plane (Knative, cert-manager, etc.)
- [ ] Implement feature serving transformer (Pinot client integration)
- [ ] Deploy embedding generation InferenceService
- [ ] Implement RAG pipeline with vector search
- [ ] Configure autoscaling policies (HPA/KEDA)
- [ ] Set up monitoring and tracing
- [ ] Network policies and security
- [ ] Documentation and runbooks

## Technical Requirements
- [ ] Kubernetes cluster (1.24+)
- [ ] Kserve installation (v0.15+)
- [ ] Apache Pinot Helm chart or manifests
- [ ] Model storage (Ceph/S3)
- [ ] Service discovery (K8s DNS or Istio)
- [ ] Observability stack (Prometheus, Grafana, Jaeger)

## Performance Targets
- **Feature serving latency**: <20ms p99 (Transformer → Pinot → Predictor)
- **Embedding generation**: <100ms p99 for 512-dim vectors
- **RAG retrieval**: <50ms p99 for top-10 vector search
- **End-to-end inference**: <200ms p99 (including Pinot queries)
- **Autoscaling**: Scale from 0 to 10 replicas in <60 seconds

## Security Considerations
- [ ] mTLS between services (Istio or service mesh)
- [ ] RBAC for Kserve InferenceServices
- [ ] Network policies (restrict Pinot access to transformers)
- [ ] Secret management (K8s secrets or Vault)
- [ ] Model signing and verification

## Monitoring and Observability
- [ ] Kserve metrics (request rate, latency, errors)
- [ ] Pinot query metrics (from transformers)
- [ ] Distributed tracing (Jaeger) across Kserve → Pinot
- [ ] Model performance metrics (accuracy, drift)
- [ ] Resource utilization dashboards

## Integration Points
- Pinot geospatial indexing (issue #9)
- Pinot vector indexing (issue #10)
- Pinot star-tree indexing (issue #11)
- Actor FSM state schema (issue #3)
- CAF-Aeron example workload (issue #1)

## Documentation Deliverables
- [ ] Kubernetes deployment guide
- [ ] Kserve + Pinot integration patterns
- [ ] Feature serving implementation guide
- [ ] RAG pipeline setup tutorial
- [ ] Monitoring and troubleshooting runbook
- [ ] Performance optimization guide

## References
- [Kserve Documentation](https://kserve.github.io/website/)
- [Kserve CNCF Announcement](https://www.cncf.io/blog/2025/11/11/kserve-becomes-a-cncf-incubating-project/)
- [Apache Pinot on Kubernetes](https://docs.pinot.apache.org/operators/operating-pinot/deployment)
- [Model Serving Patterns](https://www.kserve.io/modelserving/v1beta1/serving_runtime/)
